File: frontend/src/world/movement.ts
import type {
  MoveAnchor,
  MoveIntent,
  MoveMember,
  MovePosition,
  MovementExecution,
} from '../contracts/generated';
import type { DeepReadonly } from '../contracts/types';
import type { RuntimeSnapshot } from '../app/runtime';
import { entityRows } from './entityRows';

const radius = 6378137,
  radians = Math.PI / 180;
export const terminalExecution = (e: Pick<MovementExecution, 'state'>) =>
  ['Completed', 'Cancelled', 'Failed', 'Expired', 'Interrupted'].includes(
    e.state,
  );
export function metric(p: DeepReadonly<MoveAnchor>) {
  return [
    (p.longitudeDeg - 103.85) * radians * radius * Math.cos(1.29 * radians),
    (p.latitudeDeg - 1.29) * radians * radius,
  ];
}
export function translatedEndpoints(
  origins: readonly DeepReadonly<MovePosition>[],
  anchor: MoveAnchor,
): MovePosition[] {
  const inside = (p: DeepReadonly<MoveAnchor>) =>
    Number.isFinite(p.longitudeDeg) &&
    Number.isFinite(p.latitudeDeg) &&
    metric(p).every((v) => Math.abs(v) <= 5000);
  if (!origins.length || !inside(anchor) || origins.some((p) => !inside(p)))
    throw new Error(
      'Origins and anchor must lie within the local ±5,000 m extent.',
    );
  const points = origins.map(metric),
    a = metric(anchor);
  const center = [0, 1].map(
    (i) => points.reduce((sum, p) => sum + p[i], 0) / points.length,
  );
  const targets = points.map((p, i) => ({
    longitudeDeg: Number(
      (
        103.85 +
        (a[0] + p[0] - center[0]) /
          (radius * Math.cos(1.29 * radians) * radians)
      ).toFixed(9),
    ),
    latitudeDeg: Number(
      (1.29 + (a[1] + p[1] - center[1]) / (radius * radians)).toFixed(9),
    ),
    altitude: { ...origins[i].altitude },
  }));
  const distance = (
    p: DeepReadonly<MoveAnchor>,
    q: DeepReadonly<MoveAnchor>,
  ) => {
    const a = metric(p),
      b = metric(q);
    return Math.hypot(a[0] - b[0], a[1] - b[1]);
  };
  if (targets.some((p) => !inside(p)))
    throw new Error('A translated endpoint exceeds the local model extent.');
  if (targets.some((p, i) => distance(p, origins[i]) < 1))
    throw new Error(
      'Enter a new anchor: every member must move at least one metre.',
    );
  if (
    targets.some((p, i) =>
      targets.slice(i + 1).some((q) => distance(p, q) < 0.5),
    )
  )
    throw new Error('Coincident endpoints require explicit revision.');
  return targets;
}
export function movementReason(
  state: RuntimeSnapshot,
  id: string,
): string | undefined {
  const frame = state.presentation.frame,
    run = frame?.interactive,
    current = state.interactive.current;
  if (!frame || !run) return 'Load a local demo run.';
  const row = entityRows(frame, state.session.filters).find(
    (r) => r.entity.id === id,
  );
  if (!row) return 'Identity missing from the presented frame.';
  if (!row.visible)
    return 'Hidden by shared filters; reset filters or revise selection.';
  const controls = run.controls.filter((c) => c.entityId === id);
  if (!controls.length)
    return 'No Asset control grant; affiliation is not authority.';
  if (controls.length !== 1) return 'Ambiguous Asset/control binding.';
  const c = controls[0];
  if (!c.controlTrackId || !frame.tracks[c.controlTrackId])
    return 'No supplied control position.';
  const track = frame.tracks[c.controlTrackId];
  if (
    track.state !== 'tracking' ||
    Date.parse(frame.effectiveAt) - Date.parse(track.latest.timestamp) > 1000
  )
    return 'Control observation is stale.';
  if (
    track.latest.position.altitude.reference !== 'ELLIPSOID' ||
    track.latest.position.altitude.datumId !== 'WGS84'
  )
    return 'Control height must be ELLIPSOID/WGS84.';
  if (!c.capabilities.includes('move-horizontal'))
    return 'No horizontal movement capability.';
  if (
    row.entity.presence !== 'present' ||
    row.entity.condition !== 'operational' ||
    frame.assets[c.assetId]?.availability !== 'available'
  )
    return 'Entity or Asset is unavailable.';
  if (
    (run.executions ?? []).some(
      (e) => e.assetId === c.assetId && !terminalExecution(e),
    )
  )
    return 'Asset busy; cancel its execution and confirm termination.';
  if (state.presentation.sourceDelayed)
    return 'Source report delayed; wait for committed source recovery.';
  if (
    state.connection !== 'connected' ||
    state.presentation.status !== 'current'
  )
    return 'Reconnect for a current verified world.';
  if (
    !current ||
    current.run.executorEpoch !== run.executorEpoch ||
    current.run.lease.revision !== run.lease.revision ||
    current.run.runRevision !== run.runRevision
  )
    return 'Waiting for synchronized control status.';
  const now = state.interactive.now ?? current.serverTime;
  if (
    !current.ownsControl ||
    !run.lease.expiresAt ||
    now >= run.lease.expiresAt
  )
    return 'Acquire or explicitly reclaim control in this session.';
  if (run.state !== 'running')
    return 'Start or resume the source before moving.';
  if (
    !run.lastReportAt ||
    Date.parse(now) - Date.parse(run.lastReportAt) > 2000
  )
    return 'Source report delayed; movement requires a report within two seconds.';
  if (!c.eligible) return c.reason;
  return undefined;
}

export interface DraftParticipant {
  entityId: string;
  label: string;
  reason?: string;
  member?: MoveMember;
}
export interface MovementDraft {
  id: string;
  context: Omit<MoveIntent, 'members' | 'anchor'>;
  participants: DraftParticipant[];
  longitude: string;
  latitude: string;
  phase: 'editing' | 'reviewed' | 'submitted';
  intent?: MoveIntent;
  error?: string;
  requestId?: string;
  selectionRevision: number;
}
export function captureMovement(
  state: RuntimeSnapshot,
): MovementDraft | undefined {
  const frame = state.presentation.frame,
    run = frame?.interactive;
  if (!frame || !run) return;
  const participants = state.session.selection.items
    .filter((i) => i.kind === 'entity')
    .map(({ id }) => {
      const c = run.controls.find((c) => c.entityId === id),
        position = c?.controlTrackId
          ? frame.tracks[c.controlTrackId]?.latest.position
          : undefined;
      const member =
        c?.controlTrackId &&
        position?.altitude.reference === 'ELLIPSOID' &&
        position.altitude.datumId === 'WGS84'
          ? {
              assetId: c.assetId,
              entityId: id,
              executorId: c.executorId,
              sourceId: c.sourceId,
              controlTrackId: c.controlTrackId,
              grantId: c.grantId,
              bindingRevision: c.bindingRevision,
              busyRevision: c.busyRevision ?? 0,
              origin: structuredClone(position) as MovePosition,
              destination: structuredClone(position) as MovePosition,
            }
          : undefined;
      return {
        entityId: id,
        label: frame.entities[id]?.label ?? id,
        reason: movementReason(state, id),
        member,
      };
    });
  const origins = participants.flatMap((p) =>
    p.member ? [p.member.origin] : [],
  );
  return {
    id: crypto.randomUUID(),
    phase: 'editing',
    selectionRevision: state.session.selection.revision,
    participants,
    longitude: (origins.length
      ? origins.reduce((s, p) => s + p.longitudeDeg, 0) / origins.length
      : 103.85
    ).toFixed(9),
    latitude: (origins.length
      ? origins.reduce((s, p) => s + p.latitudeDeg, 0) / origins.length
      : 1.29
    ).toFixed(9),
    context: {
      missionId: frame.mission.id,
      runId: run.runId,
      executorEpoch: run.executorEpoch,
      sourceId: run.sourceId,
      grantId: run.grantId,
      grantRevision: run.grantRevision,
      reviewedFrameId: frame.frameId,
      modelId: 'local-horizontal-v1',
      deadline: new Date(Date.parse(frame.recordedAt) + 30_000).toISOString(),
    },
  };
}
export function reviewMovement(draft: DeepReadonly<MovementDraft>): MoveIntent {
  if (
    !draft.participants.length ||
    draft.participants.some((p) => p.reason || !p.member)
  )
    throw new Error(
      'Every frozen participant must be eligible. Revise selection and create a new draft.',
    );
  if (!draft.longitude.trim() || !draft.latitude.trim())
    throw new Error('Enter both longitude and latitude.');
  const members = draft.participants.map((p) =>
    structuredClone(p.member!),
  ) as MoveMember[];
  const anchor = {
    longitudeDeg: Number(draft.longitude),
    latitudeDeg: Number(draft.latitude),
  };
  const targets = translatedEndpoints(
    members.map((m) => m.origin),
    anchor,
  );
  return {
    ...draft.context,
    anchor,
    members: members.map((m, i) => ({
      ...m,
      destination: targets[i],
    })) as MoveIntent['members'],
  };
}
export function draftReason(state: RuntimeSnapshot): string | undefined {
  const d = state.session.movementDraft;
  if (!d) return 'Create a draft from the selected members.';
  if (d.phase !== 'reviewed' || !d.intent)
    return 'Review every endpoint before submission.';
  if (state.interactive.pending || state.interactive.busy)
    return 'Wait or reconcile the pending request.';
  const now = state.interactive.now ?? state.interactive.current?.serverTime;
  if (!now || now >= d.context.deadline)
    return 'Review expired; create a fresh draft explicitly.';
  const run = state.presentation.frame?.interactive;
  if (
    !run ||
    run.executorEpoch !== d.context.executorEpoch ||
    run.grantRevision !== d.context.grantRevision ||
    run.runId !== d.context.runId
  )
    return 'Run or grant changed. Create a fresh draft explicitly.';
  for (const p of d.participants) {
    const reason = p.reason ?? movementReason(state, p.entityId);
    if (reason) return `${p.label}: ${reason}`;
    const control = state.presentation.frame?.interactive?.controls.find(
      (c) => c.assetId === p.member?.assetId,
    );
    if (
      !control ||
      control.bindingRevision !== p.member?.bindingRevision ||
      (control.busyRevision ?? 0) !== p.member?.busyRevision
    )
      return `${p.label}: binding or reservation changed. Create a new draft.`;
  }
  return undefined;
}

File: backend/app/commands/movement.py
"""Candidate movement transitions, composed by InteractiveService's single writer."""
from datetime import datetime
from app.commands.kinematics import step
from app.commands.zone_rules import blocked

TERMINAL = {"Completed", "Cancelled", "Failed", "Expired", "Interrupted"}


def age(newer, older):
    return (datetime.fromisoformat(newer.replace("Z", "+00:00")) - datetime.fromisoformat(older.replace("Z", "+00:00"))).total_seconds()


def controls_for(frame, entity_id):
    return [c for c in frame["interactive"]["controls"] if c["entityId"] == entity_id]


def control_reason(frame, control, now, ignore_busy=False):
    run = frame["interactive"]
    if run["state"] != "running":
        return "INVALID_TRANSITION", "Source must be running."
    if not run.get("lastReportAt") or age(now, run["lastReportAt"]) > 2:
        return "SOURCE_UNHEALTHY", "No current source report within two seconds."
    if len(controls_for(frame, control["entityId"])) != 1:
        return "SELECTION_INVALID", "Ambiguous Asset/control binding."
    if "move-horizontal" not in control["capabilities"]:
        return "CONTROL_REQUIRED", "No horizontal movement capability."
    entity, asset = frame["entities"].get(control["entityId"]), frame["assets"].get(control["assetId"])
    if not entity or entity["presence"] != "present" or entity["condition"] != "operational" or not asset or asset["availability"] != "available":
        return "SELECTION_INVALID", "Entity or Asset is unavailable for movement."
    track = frame["tracks"].get(control.get("controlTrackId"))
    if not track or track["state"] != "tracking" or age(frame["effectiveAt"], track["latest"]["timestamp"]) > 1:
        return "POSITION_UNAVAILABLE", "Control position is missing or stale."
    if track["source"]["id"] != control["sourceId"] or track["entityId"] != control["entityId"]:
        return "BINDING_CHANGED", "Control Track/source binding changed."
    altitude = track["latest"]["position"]["altitude"]
    if altitude["reference"] != "ELLIPSOID" or altitude.get("datumId") != "WGS84":
        return "UNSUPPORTED_REFERENCE", "Movement requires supplied ELLIPSOID/WGS84 height."
    zone_reason = blocked(frame, track["latest"]["position"])
    if zone_reason:
        return "ENDPOINT_INVALID", zone_reason
    if not ignore_busy and any(e["assetId"] == control["assetId"] and e["state"] not in TERMINAL for e in run.get("executions", [])):
        return "ASSET_BUSY", "Asset is busy; Cancel its execution and wait for termination."
    return None


def direct_member_reason(frame, control):
    """Availability is evidence, not affiliation or the existence of a display Track."""
    entity = frame["entities"].get(control["entityId"])
    asset = frame["assets"].get(control["assetId"])
    if not entity or entity["presence"] == "removed":
        return "UNAVAILABLE", "Entity is no longer present."
    if entity["condition"] == "non-operational":
        return "UNAVAILABLE", "Down: reported non-operational."
    if entity["presence"] != "present":
        return "NO_RESPONSE", "No response."
    track = frame["tracks"].get(control.get("controlTrackId"))
    if not track:
        return "POSITION_UNAVAILABLE", "No known control position."
    if track["state"] != "tracking" or age(frame["effectiveAt"], track["latest"]["timestamp"]) > 1:
        return "NO_RESPONSE", "No response: control observation is stale."
    altitude = track["latest"]["position"]["altitude"]
    if altitude["reference"] != "ELLIPSOID" or altitude.get("datumId") != "WGS84":
        return "UNSUPPORTED_REFERENCE", "Unsupported altitude reference."
    if entity["condition"] != "operational" or not asset or asset["availability"] != "available" or "move-horizontal" not in control["capabilities"]:
        return "UNAVAILABLE", "Unavailable for movement."
    return None


def transition(execution, state, reason=None):
    if execution["state"] != state:
        execution["revision"] += 1
    execution["state"] = state
    if reason:
        execution["reason"] = reason


def terminate(frame, execution, state, reason):
    if execution["state"] in TERMINAL:
        return
    transition(execution, state, reason)
    execution["terminalSequence"] = frame["sequence"] + 1
    control = next((c for c in frame["interactive"]["controls"] if c["assetId"] == execution["assetId"]), None)
    if control:
        control["busyRevision"] += 1
    track = frame["tracks"].get(execution["controlTrackId"])
    if track:
        track["latest"]["velocity"] = {"speedMps": 0.0, "headingTrueDeg": track["latest"].get("velocity", {}).get("headingTrueDeg", 0.0)}


def project(frame, checkpoint, now):
    bound_assets = {c["assetId"] for c in frame["interactive"]["controls"]}
    checkpoint["directOrders"] = {asset_id: value for asset_id, value in checkpoint.get("directOrders", {}).items()
                                 if asset_id in bound_assets}
    executions = [e for e in checkpoint["executions"] if e.get("kind") == "move"]
    active = [e for e in executions if e["state"] not in TERMINAL]
    terminal = sorted((e for e in executions if e["state"] in TERMINAL), key=lambda e: (e["terminalSequence"], e["acceptedSequence"], e["id"]))
    retained = active + terminal[-(64 - len(active)):]
    checkpoint["executions"] = retained + [e for e in checkpoint["executions"] if e.get("kind") != "move"]
    checkpoint["schemaVersion"] = "1.3"
    frame["interactive"]["executions"] = retained
    if "cancel" not in frame["interactive"]["supportedActions"]:
        frame["interactive"]["supportedActions"].append("cancel")
    for c in frame["interactive"]["controls"]:
        c.setdefault("busyRevision", 0)
        prior = checkpoint.get("directOrders", {}).get(c["assetId"])
        if prior:
            c["lastDirectOrder"] = prior
        track = frame['tracks'].get(c.get('controlTrackId'))
        zone_reason = blocked(frame, track['latest']['position']) if track else None
        reason = ("ENDPOINT_INVALID", zone_reason) if zone_reason else control_reason(frame, c, now, ignore_busy=True)
        c.update(eligible=reason is None, reason=reason[1] if reason else "Eligible for horizontal movement.")


def advance(frame, checkpoint, now):
    """Called even when paused: initial-execution expiry uses elapsed backend time."""
    events = []
    run = frame["interactive"]
    for e in checkpoint["executions"]:
        if e.get("kind") != "move" or e["state"] in TERMINAL:
            continue
        before = e["state"]
        control = next((c for c in run["controls"] if c["assetId"] == e["assetId"]), None)
        if not e.get("startedAt") and now >= e["deadline"]:
            terminate(frame, e, "Expired", "Initial execution deadline elapsed.")
        elif run["state"] == "running":
            reason = control_reason(frame, control, now, True) if control else ("BINDING_CHANGED", "Control binding missing.")
            if e["executorEpoch"] != run["executorEpoch"]:
                terminate(frame, e, "Interrupted", "Executor epoch changed.")
            elif e["grantRevision"] != run["grantRevision"]:
                terminate(frame, e, "Cancelled", "Capability explicitly revoked.")
            elif reason or not control or (e["bindingRevision"], e["reservationRevision"], e["controlTrackId"], e["sourceId"]) != (control["bindingRevision"], control["busyRevision"], control.get("controlTrackId"), control["sourceId"]):
                terminate(frame, e, "Failed", reason[1] if reason else "Binding or reservation changed.")
            else:
                if not e.get("startedAt"):
                    e.update(startedAt=frame["effectiveAt"], startedTick=run["tick"])
                transition(e, "Running")
                candidate = dict(e)
                position, velocity = step(candidate)
                track = frame["tracks"][e["controlTrackId"]]
                zone_reason = blocked(frame, track["latest"]["position"], position)
                if zone_reason:
                    terminate(frame, e, "Failed", zone_reason)
                else:
                    e.update(candidate)
                    track["latest"].update(position=position, velocity=velocity)
                if not zone_reason and e["remainingMetres"] == 0:
                    terminate(frame, e, "Completed", "Destination reached in a committed source sample.")
                    e["completionSample"] = {"sequence": frame["sequence"] + 1, "trackId": e["controlTrackId"], "timestamp": frame["effectiveAt"], "position": position}
        if e["state"] != before:
            events.append({"effectiveAt": frame["effectiveAt"], "type": f'movement.{e["state"].lower()}', "severity": "info",
                           "entityIds": [e["entityId"]], "source": {"id": run["sourceId"], "kind": "simulation", "mode": "simulated"},
                           "extensions": {"sentinel.movement": {"executionId": e["id"], "commandId": e["commandId"], "reason": e.get("reason")}}})
    return events
